packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,283 @@
+  const language = contentSource.currentLanguage;
+  const content = useMemo(() => {
+    return createContent(contentSource);
+  }, [language]);
+  const requiredConsents = content.requiredCookies
+    ? getConsentsFromCookieGroup(content.requiredCookies.groups)
+    : undefined;
+  const optionalConsents = content.optionalCookies
+    ? getConsentsFromCookieGroup(content.optionalCookies.groups)
+    : undefined;
+  const consentController = useMemo(
+    () => createConsentController({ requiredConsents, optionalConsents, cookieDomain }),
+    [],
+  );
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const notifyOnAllConsentsGiven = () => {
+    if (content.onAllConsentsGiven && hasUserHandledAllConsents()) {
+      content.onAllConsentsGiven(mergeConsents());
+    }
+    if (content.focusTargetSelector) {
+      forceFocusToElement(content.focusTargetSelector);
+    }
+  };
+
+  const setOptional = (approved: boolean) => {
+    Object.keys(consentController.getOptional()).forEach((optionalConsent) => {
+      consentController.update(optionalConsent, approved);
+    });
+  };
+  const deselectOptional = () => {
+    setOptional(false);
+  };
+  const selectOptional = () => {
+    setOptional(true);
+  };
+  const approveRequiredAndSaveAllGivenConsents = () => {
+    consentController.approveRequired();
+    consentController.save();
+    notifyOnAllConsentsGiven();
+  };
+  const approveAllAndSave: ConsentController['approveAll'] = () => {
+    consentController.approveAll();
+    consentController.save();
+    notifyOnAllConsentsGiven();
+  };
+  const approveOnlyRequiredAndSave: ConsentController['approveRequired'] = () => {
+    deselectOptional();
+    consentController.approveRequired();
+    consentController.save();
+    notifyOnAllConsentsGiven();
+  };
+
+  const onAction: CookieConsentContextType['onAction'] = (action, consents, value) => {
+    if (action === 'approveAll') {
+      approveAllAndSave();
+    } else if (action === 'approveRequired') {
+      approveOnlyRequiredAndSave();
+    } else if (action === 'approveSelectedAndRequired') {
+      approveRequiredAndSaveAllGivenConsents();
+    } else if (action === 'changeCookieGroupConsents') {
+      consents.forEach((consent) => {
+        consentController.update(consent, value);
+      });
+    } else if (action === 'approveOptional') {
+      selectOptional();
+    } else if (action === 'unapproveOptional') {
+      deselectOptional();
+    }
+    reRender();
+  };
+
+  const countApprovedOptional: CookieConsentContextType['countApprovedOptional'] = () => {
```

Comment:
Would `getApproveRatioForOptional` be more descriptive? Or `getApprovalRatingForOptional`?

Reply:
Changed to getApprovalPercentageForOptional